### ISSUE TITLE
clang rejects valid code that uses _Atomic

### ISSUE THREAD

**Issue by 021de1a4-aab3-4b45-ada0-9b4a8d155441** [llvm/llvm-project#48742]
|  |  |
| --- | --- |
| Bugzilla Link | [49398](https://llvm.org/bz49398) |
| Version | trunk |
| OS | Linux |
| Attachments | [Valid C program that clang rejects](https://user-images.githubusercontent.com/92601859/143761849-b36fc870-f63d-4868-8801-40552f70fa3a.gz) |
| CC | @DougGregor,@zygoloid |

## Extended Description 

Command (where ex.c is the attached file):

clang -std=c17 -c ex.c

LLVM revision from which clang was built: 493f1407927cdf6db2e08f150ed9f9b8abe4ac45

Expected behaviour: the program should compile successfully.

Actual behaviour: clang gives this error:

ex.c:3:7: error: initializing 'int' with an expression of incompatible type '_Atomic(int)'

The program is successfully compiled by gcc 9.3.0.

The program in the attached file is:

void foo() {
  _Atomic int x;
  int y = (x = 2);
}

which I believe is valid: it's fine to initialize a non-atomic int using an atomic int.

If the declaration of y is changed to:

  int y = x;

then the program compiles successfully.

**Comment by AaronBallman** [comment]
It looks like Clang is confused as to the type `(x = 2)` has. According to C2x 6.5.16p3: "The type of an assignment expression is the type the left operand would have after lvalue conversion." 6.3.2.1p2 says: "...; additionally, if the lvalue has atomic type, the value has the non-atomic version of the type of the lvalue". So the resulting type should be `int` and not `_Atomic(int)`. This is why we fail this: 
```
void foo() {
  _Atomic int x;
  _Static_assert(__builtin_types_compatible_p(__typeof__(x = 2), int), "incorrect");
}
```
https://godbolt.org/z/jfMKadbhW

**Comment by AaronBallman** [comment]
https://reviews.llvm.org/D122496 addresses this.

**Comment by AaronBallman** [comment]
This should now be fixed in bfa2f25d350c1015b74b8a14684e68efa6500bbc.
